test_libs/pyspec/eth2spec/test/phase_0/epoch_processing/test_process_justification_and_finalization.py

```diff
@@ -0,0 +1,273 @@
+    state.justification_bits[1:3] = [1, 1]  # mock 3rd and 4th latest epochs as justified (indices are pre-shift)
+    # mock the 2nd latest epoch as justifiable, with 4th as source
+    add_mock_attestations(spec, state,
+                          epoch=epoch - 2,
+                          source=c4,
+                          target=c2,
+                          sufficient_support=sufficient_support)
+
+    # process!
+    yield from run_process_just_and_fin(spec, state)
+
+    assert state.previous_justified_checkpoint == c3  # changed to old current
+    if sufficient_support:
+        assert state.current_justified_checkpoint == c2  # changed to 2nd latest
+        assert state.finalized_checkpoint == c4  # finalized old previous justified epoch
+    else:
+        assert state.current_justified_checkpoint == c3  # still old current
+        assert state.finalized_checkpoint == old_finalized  # no new finalized
+
+
+def finalize_on_23(spec, state, epoch, sufficient_support):
+    assert epoch > 3
+    state.slot = (spec.SLOTS_PER_EPOCH * epoch) - 1  # skip ahead to just before epoch
+
+    # 43210 -- epochs ago
+    # 210xx  -- justification bitfield indices (pre shift)
+    # 3210x -- justification bitfield indices (post shift)
+    # 01*0. -- justification bitfield contents, . = this epoch, * is being justified now
+    # checkpoints for the epochs ago:
+    c3 = spec.Checkpoint(epoch=epoch - 3, root=b'\xaa' * 32)
+    c2 = spec.Checkpoint(epoch=epoch - 2, root=b'\xbb' * 32)
+    state.block_roots[spec.get_epoch_start_slot(c3.epoch) % spec.SLOTS_PER_HISTORICAL_ROOT] = c3.root
+    state.block_roots[spec.get_epoch_start_slot(c2.epoch) % spec.SLOTS_PER_HISTORICAL_ROOT] = c2.root
+
+    old_finalized = state.finalized_checkpoint
+    state.previous_justified_checkpoint = c3
+    state.current_justified_checkpoint = c3
+    state.justification_bits = spec.Bitvector[spec.JUSTIFICATION_BITS_LENGTH]()
+    state.justification_bits[1] = 1  # mock 3rd latest epoch as justified (indices are pre-shift)
+    # mock the 2nd latest epoch as justifiable, with 3rd as source
+    add_mock_attestations(spec, state,
+                          epoch=epoch - 2,
+                          source=c3,
+                          target=c2,
+                          sufficient_support=sufficient_support)
+
+    # process!
+    yield from run_process_just_and_fin(spec, state)
+
+    assert state.previous_justified_checkpoint == c3  # changed to old current
+    if sufficient_support:
+        assert state.current_justified_checkpoint == c2  # changed to 2nd latest
+        assert state.finalized_checkpoint == c3  # finalized old previous justified epoch
+    else:
+        assert state.current_justified_checkpoint == c3  # still old current
+        assert state.finalized_checkpoint == old_finalized  # no new finalized
+
+
+def finalize_on_123(spec, state, epoch, sufficient_support):
```

Comment:
This scenario isn't really correct in how it would manifest on chain.

* Need to have current_justified as `c3` and previous_justified as `c5`
* need to add mock attestations for `(epoch - 2, c5, c2)` and for `(epoch - 1, c3, c1)`
* justification_bits just set to `bits[1] = 1`
* This will justify `c2` and `c1` and will finalize `c3` using `(c5 -> c2)` as the center of the `(c3 -> c1)` link.

Reply:
See 36dd977